File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def bfs(maps, startX, startY):

    # 왼쪽, 오른쪽, 아래, 위 -> y가 클수록 아래 줄임에 유의
    dx = [-1,1,0,0]
    dy = [0,0,1,-1]
    
    n = len(maps)
    m = len(maps[0])

    dq = deque()
    dq.append((startX, startY))    

    while dq:
        x, y = dq.popleft()

        for i in range(4): # 4 방향 검사
            nx = x + dx[i]
            ny = y + dy[i]

            if nx == startX and ny == startY: # 시작 위치인가?
                continue 

            if nx < 0 or nx >= n or ny < 0 or ny >= m: # 영역 내 검사
                continue

            if maps[nx][ny] == 0: # 벽 검사
                continue

            if maps[nx][ny] == 1:
                maps[nx][ny] = maps[x][y] + 1
                dq.append((nx, ny))

    goal = maps[- 1][- 1]    
    return goal if goal != 1 else -1    


def solution(maps):    
    return bfs(maps, 0,0)

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.cpp
#include<vector>
#include <deque>
using namespace std;

int bfs(vector<vector<int> > maps, int startX, int startY)
{
    vector<int> dx = {-1,1,0,0};
    vector<int> dy = {0,0,1,-1};
    
    int n = maps.size();
    int m = maps[0].size();
    
    deque<int> dqX, dqY;
    dqX.push_back(startX);
    dqY.push_back(startY);
    
    while(dqX.size() > 0)
    {
        int x = dqX.front(); 
        int y = dqY.front();
        dqX.pop_front();
        dqY.pop_front();
        
        for(int i = 0; i< 4; i++)
        {
            int nx = x + dx[i];
            int ny = y + dy[i];
            
            if(nx == startX && ny == startY) continue;
            if( nx < 0 || nx >= n || ny < 0 || ny >= m) continue;
            if(maps[nx][ny] == 0) continue;
            
            if(maps[nx][ny] == 1)
            {
                maps[nx][ny] = maps[x][y] + 1;
                dqX.push_back(nx);
                dqY.push_back(ny);
            }
        }        
    }
    
    int goal = maps[n-1][m-1];
    if(goal != 1) return goal;
    else return -1;    
}

int solution(vector<vector<int> > maps)
{
    return bfs(maps, 0,0);
}